### PR TITLE
PAT-393: Assignment of controls, interventions, actions and assessments as part of the CIM

### PR DESCRIPTION
**Pull request by timharrison-moj**
* Allocates controls, actions, interventions, assessments, mandatory or optional, based on config rules.
* Rules use the offender's tact status, risk band, needs answers, gender and location.
* These will be used by the Manage Interventions form as the starting point for C&I.
* The rules governing whether each C&I is mandatory, optional or not suitable are covered by 38 unit tests.

### CODE REVIEW

File: server/config/controlsAndInterventions/actions/getActions.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  // Get a list of the unique needs attributes from the needs priority form data
+  const needsAttributes = R.flatten(getNeedsAttributesFromForm(needsForm))
+
+  // Traverse the actions to match them with this offender
+  allActions.forEach(a => {
+    const action = actionMatcher(needsAttributes, a, gender, riskBand, tactOffender, custodyOrCommunity)
+    if (!isNilOrEmpty(action)) {
+      actions.push(action)
+    }
+  })
```
Comment (steverendell) [inline server/config/controlsAndInterventions/actions/getActions.js]: i'd suggest doing this in a pipeline, i.e
const actions = allActions
  .map(a => actionMatcher(needsAttributes, a, gender, riskBand, tactOffender, custodyOrCommunity)
  .filter (a => !isNilOrEmpty(a))
(i think!)

File: server/config/controlsAndInterventions/actions/getActions.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+   */
+
+  const match = action.rules.some(r => {
+    return needsAttributes.includes(r.attribute) && genderMatch(r.gender, gender) && tactRuleMatch(r.tact, tactOffender)
+  })
+
+  return match === true
+    ? { uid: action.uid, desc: action.desc, mandatory: true }
+    : { uid: action.uid, desc: action.desc, mandatory: false }
+}
```
Comment (steverendell) [inline server/config/controlsAndInterventions/actions/getActions.js]: the ternary operator looks redundant here

File: server/config/controlsAndInterventions/assessments/getAssessments.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * @returns { assessments : List { uid, desc, mandatory }}
+ */
+const getAssessments = (nominalId, needsForm, riskBand, gender, tactOffender, custodyOrCommunity) => {
+  const assessments = []
+
+  // Get a list of the unique needs attributes from the needs priority form data
+  const needsAttributes = R.flatten(getNeedsAttributesFromForm(needsForm))
+
+  // Traverse the assessments and match with this offender
```
Comment (steverendell) [inline server/config/controlsAndInterventions/assessments/getAssessments.js]: similar to comment above

File: server/config/controlsAndInterventions/assessments/getAssessments.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+   * Attempt to match on the rules for this assessment.
+   * There are no reverse rules currently defined for assessments (as there are for interventions).
+   * The gender and tact status also apply at the rule level, as well as the assessment level.
+   */
+
+  const match = assessment.rules.some(r => {
+    return needsAttributes.includes(r.attribute) && genderMatch(r.gender, gender) && tactRuleMatch(r.tact, tactOffender)
+  })
+
+  return match === true
```
Comment (steverendell) [inline server/config/controlsAndInterventions/assessments/getAssessments.js]: similar to comment above

File: server/config/controlsAndInterventions/common/utils.js [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+const mandatoryBandMatch = (mandatoryBands, riskBand) => {
+  if (isNilOrEmpty(mandatoryBands) || mandatoryBands === 'NA') {
+    return false
+  }
+  const bands = mandatoryBands.split(',')
+  return bands.includes(riskBand.toString())
+}
+
+const getNeedsAttributesFromForm = needsForm => {
```
Comment (steverendell) [inline server/config/controlsAndInterventions/common/utils.js]: not sure here but might be simpler to use ramda to find the properties you want

File: server/config/controlsAndInterventions/controls/getControls.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+ *
+ * @returns { controls : List { uid, desc, mandatory }}
+ */
+const getControls = (nominalId, needsForm, riskBand, gender, tactOffender, custodyOrCommunity) => {
+  const controls = []
+
+  // Get a list of the unique needs attributes from the needs priority form data
+  const needsAttributes = R.flatten(getNeedsAttributesFromForm(needsForm))
+
+  // Traverse the controls to match with this offender
```
Comment (steverendell) [inline server/config/controlsAndInterventions/controls/getControls.js]: as above

File: server/config/controlsAndInterventions/controls/getControls.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+  /*
+   * Attempt to match based on the rules for this control (though no rules currently exist for controls)
+   * The gender and tact status also apply at the rule level, as well as the control level.
+   */
+
+  const match = control.rules.some(r => {
+    return needsAttributes.includes(r.attribute) && genderMatch(r.gender, gender) && tactRuleMatch(r.tact, tactOffender)
+  })
+
```
Comment (steverendell) [inline server/config/controlsAndInterventions/controls/getControls.js]: as above

File: server/config/controlsAndInterventions/interventions/getInterventions.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+ * @param custodyOrCommunity Contains 'community' or 'custody' based on offender location
+ *
+ * @returns { interventions : List { uid, desc, mandatory }}
+ */
+const getInterventions = (nominalId, needsForm, riskBand, gender, tactOffender, custodyOrCommunity) => {
+  const interventions = []
+
+  // Get a list of the unique needs attributes from the needs priority form data
+  const needsAttributes = R.flatten(getNeedsAttributesFromForm(needsForm))
+
```
Comment (steverendell) [inline server/config/controlsAndInterventions/interventions/getInterventions.js]: as above

File: server/config/controlsAndInterventions/interventions/getInterventions.js [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+      'reverse' in r &&
+      r.reverse === true &&
+      !needsAttributes.includes(r.attribute) &&
+      genderMatch(r.gender, gender) &&
+      tactRuleMatch(r.tact, tactOffender)
+
+    return matchedStandardRule || matchedReverseRule
+  })
+
+  return match === true
```
Comment (steverendell) [inline server/config/controlsAndInterventions/interventions/getInterventions.js]: as above

File: server/config/controlsAndInterventions/actions/getActions.test.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
+const R = require('ramda')
+const actions = require('./getActions')
+
```
Comment (steverendell) [inline server/config/controlsAndInterventions/actions/getActions.test.js]: I dont think any of your tests need to be async

Comment (timharrison-moj) [reply server/config/controlsAndInterventions/actions/getActions.test.js]: Agreed. Removing the async - missed that.

File: server/config/controlsAndInterventions/common/utils.js (hunk 5 above)
Comment (timharrison-moj) [reply server/config/controlsAndInterventions/common/utils.js]: I did look ramda functions but didn't find a simpler way. The current way is probably clear enough - it recursively picks out all keys called 'needsAttribute' from a document and grabs the values associated. 8 lines of code that does the job well.

File: server/config/controlsAndInterventions/actions/getActions.js (hunk 1 above)
Comment (timharrison-moj) [reply server/config/controlsAndInterventions/actions/getActions.js]: Yes, agreed. Just tried it for actions and reduces the code by a few lines. Will apply to the four places this is used, thanks.

File: server/config/controlsAndInterventions/actions/getActions.js (hunk 2 above)
Comment (timharrison-moj) [reply server/config/controlsAndInterventions/actions/getActions.js]: You're right - is been refined and altered a couple of times, and now the return values are so similar that we can do this in a single line. I will change this.

File: server/config/controlsAndInterventions/assessments/getAssessments.js (hunk 4 above)
Comment (timharrison-moj) [reply server/config/controlsAndInterventions/assessments/getAssessments.js]: Agreed.

File: server/config/controlsAndInterventions/controls/getControls.js (hunk 6 above)
Comment (timharrison-moj) [reply server/config/controlsAndInterventions/controls/getControls.js]: And resolved as above.

File: server/config/controlsAndInterventions/controls/getControls.js (hunk 7 above)
Comment (timharrison-moj) [reply server/config/controlsAndInterventions/controls/getControls.js]: And resolved as above

File: server/config/controlsAndInterventions/interventions/getInterventions.js (hunk 8 above)
Comment (timharrison-moj) [reply server/config/controlsAndInterventions/interventions/getInterventions.js]: And resolved as above

File: server/config/controlsAndInterventions/interventions/getInterventions.js (hunk 9 above)
Comment (timharrison-moj) [reply server/config/controlsAndInterventions/interventions/getInterventions.js]: Resolved.

File: server/config/controlsAndInterventions/assessments/getAssessments.js (hunk 3 above)
Comment (timharrison-moj) [reply server/config/controlsAndInterventions/assessments/getAssessments.js]: Agrred

**Pull request merged**